ast.c

```diff
@@ -724,6 +720,53 @@ static void ast_to_zval(zval *zv, zend_ast *ast, ast_state_info_t *state) {
 			ast->kind = ZEND_AST_ASSIGN_OP;
 			ast->attr = AST_BINARY_COALESCE;
 			break;
+		case ZEND_AST_CLASS_NAME:
+			if (state->version < 70) {
+				zval name_zval;
+				ast_to_zval(&name_zval, ast->child[0], state);
+				zval class_name_zval;
+				ast_create_virtual_node_ex(
+						&class_name_zval, AST_NAME, ast->child[0]->attr, zend_ast_get_lineno(ast), state, 1, &name_zval);
+				zval const_zval;
+				ZVAL_STR_COPY(&const_zval, AST_STR(str_class));
+				ast_create_virtual_node_ex(
+						zv, ZEND_AST_CLASS_CONST, 0, zend_ast_get_lineno(ast), state, 2, &class_name_zval, &const_zval);
+				return;
+			}
+			break;
+#else
+		case ZEND_AST_CLASS_CONST:
+			if (state->version >= 70) {
+				// Convert to an AST_CLASS_NAME instead. This is the opposite of the work done in the ZEND_AST_CLASS_NAME case.
+				zend_ast *const_name_ast = ast->child[1];
+				if (const_name_ast->kind != ZEND_AST_ZVAL) {
+					break; // e.g. (new stdClass())::class is a runtime error, not a parse error
+				}
+				zval const_name_zv = *zend_ast_get_zval(const_name_ast);
+				if (Z_TYPE(const_name_zv) != IS_STRING) {
+					break; // e.g. (2)::class (with brackets)
+				}
+				zend_string *str = Z_STR(const_name_zv);
+				if (zend_string_equals_ci(AST_STR(str_class), str)) {
+					zend_ast *class_name_ast = ast->child[0];
+					zval class_name_zval;
+					if (class_name_ast->kind == ZEND_AST_ZVAL) {
+						// e.g. Foo::class
+						zval class_name_raw_zval;
+						ZVAL_COPY(&class_name_raw_zval, zend_ast_get_zval(class_name_ast));
+						ast_create_virtual_node_ex(
+								&class_name_zval, AST_NAME, class_name_ast->attr, zend_ast_get_lineno(class_name_ast), state, 1, &class_name_raw_zval);
+					} else {
+						// e.g. []::class (not a parse error, but a runtime error)
+						ast_to_zval(&class_name_zval, class_name_ast, state);
+					}
```

Comment:
Wouldn't `ast_to_zval` here work for both branches?

Reply:
It didn't create the `ast\Node` when I tried that, probably because the parent kind or position was wrong.

Reply:
Ah yeah, I guess it will just convert it to a simple value. In that case it's fine.